isabelle/Merkle/MerkleInterpreter.thy

```diff
@@ -0,0 +1,255 @@
+theory MerkleInterpreter
+
+imports MerkleTypes Core.Semantics
+begin
+
+section "Interpreter"
+
+fun getAction :: "MCase \<Rightarrow> Action" where
+  "getAction (Case action _) = action"
+| "getAction (MerkleizedCase action _) = action"
+
+fun getInputContent :: "MInput \<Rightarrow> Input" where 
+  "getInputContent (NormalInput i) = i" 
+| "getInputContent (MerkleizedInput i _ _) = i"
+
+fun getContinuation :: "MInput \<Rightarrow> MCase \<Rightarrow> MContract option" where
+  "getContinuation (NormalInput _) (Case _ continuation) = Some continuation" 
+| "getContinuation (MerkleizedInput _ inputContinuationHash continuation)  (MerkleizedCase _ continuationHash)
+    = (if inputContinuationHash = continuationHash 
+       then Some continuation 
+       else None
+      )" 
+| "getContinuation _ _ = None"
+
+subsection "Reduce contract Step"
+
+\<comment> \<open>The only difference with ReduceStepResult is the MContract instead of Contract\<close>
+datatype ReduceStepMResult = Reduced ReduceWarning ReduceEffect State MContract
+                          | NotReduced
+                          | AmbiguousTimeIntervalReductionError
+
+fun reduceMContractStep :: "Environment \<Rightarrow> State \<Rightarrow> MContract \<Rightarrow> ReduceStepMResult" where
```

Comment:
To improve readability in the future I might rename the functions and type to be the same as the Semantic module, and whenever we need to disambiguate we use the qualified name